### PR TITLE
Sphere Fitting  local estimator

### PR DESCRIPTION
**Pull request by dcoeurjo**
- New SphereFitting local functor [Mellado et al., SGP 2012]
- fixes couple of issues with new CSurfelLocal concept.

this PR also contains EstimatorCache commits.

### CODE REVIEW

**Comment by dcoeurjo** [comment]
Can someone review this PR ?

**Comment by kerautret** [comment]
yes look it

File: ChangeLog.md [hunk 1]
```diff
@@ -8,6 +8,11 @@
  - New EstimatorCache class to cache quantities estimated by a
    surfel local estimator. (David Coeurjolly,
    [#927](https://github.com/DGtal-team/DGtal/pull/927))
+ - New digital surface local estimator that compute a sphere
```
Comment (kerautret) [inline ChangeLog.md]: that computeS

File: cmake/CheckDGtalOptionalDependencies.cmake [hunk 2]
```diff
@@ -432,6 +440,30 @@ IF(WITH_CGAL)
...
+    SET(DGtalLibInc ${DGtalLibInc} ${PATATE_INCLUDE_DIR})
+    INCLUDE_DIRECTORIES( ${PATATE_INCLUDE_DIR})
+    SET(PATATE_FOUND_DGTAL 1)
+    ADD_DEFINITIONS("-DWITH_PATATE ")
+    ADD_DEFINITIONS("-DWITH_Eigen3 ")
+    message(STATUS "PATATE found. ${PATATE_INCLUDE_DIR} ")
+ ELSE(PATATE_FOUND)
+   message(FATAL_ERROR "Patate headers not found.")
+ ENDIF(PATATE_FOUND)
+ENDIF(WITH_PATATE)
```
Comment (kerautret) [inline cmake/CheckDGtalOptionalDependencies.cmake]: Cmake Patate integration look fine

File: examples/geometry/surfaces/exampleEstimatorFromSurfelFunctors.cpp [hunk 3]
```diff
@@ -117,7 +117,7 @@ int main(  )
```
Comment (kerautret) [inline examples/geometry/surfaces/exampleEstimatorFromSurfelFunctors.cpp]: I have an error but not related to this PR: 
[ERR] Assertion Error - assertion (( false && "[ConstAlias::ConstAlias( T&& )] Const-aliasing a rvalue ref has no meaning. Consider Clone instead." )) failed in DGtal::ConstAlias<DGtal::CanonicSCellEmbedder<DGtal::KhalimskySpaceND<3, int> > >::ConstAlias(T &&) [T = DGtal::CanonicSCellEmbedder<DGtal::KhalimskySpaceND<3, int> >]: /Users/kerautre/EnCours/DGtal/src/DGtal/base/ConstAlias.h(259)

File: src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+     * \brief Aim: Use patate library to perform a local sphere fitting.
+     *
+     * model of CLocalEstimatorFromSurfelFunctor.
+     *
+     *
+     * @tparam TSurfel type of surfels
+     * @tparam TEmbedder type of functors which embed surfel to @f$
+     * \mathbb{R}^3@f$.
+     * @tparam TNormalVectorEstimatorCache the type of normal vector
+     * czche to consider (see EstimatorCache class).
```
Comment (kerautret) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: czche

Comment (kerautret) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: Perhaps mention an example where it is used ? (tests or example)

File: src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+       * @param h gridstep.
+       */
+      SphereFittingEstimator(ConstAlias<SCellEmbedder> anEmbedder,
+                             const double h,
+                             ConstAlias<NormalVectorEstimatorCache> anEstimator):
+        myEmbedder(&anEmbedder), myH(h), myNormalEsitmatorCache(&anEstimator)
+      {
+        //From Mellado's example
+        myFit = new Fit();
+        myFit->setWeightFunc(WeightFunc(100.0));
```
Comment (kerautret) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: the weight cannot be changed ? (perhaps add an optional parameter with default value in the constructor )

File: src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+      /**
+       * Reset the point list.
+       *
+       */
+      void reset()
+      {
+        delete myFit;
+        myFit = new Fit();
+        myFit->setWeightFunc(WeightFunc(100.0));
```
Comment (kerautret) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: also here perhaps add a constant even if the weight has no need to be changed

File: src/DGtal/io/Display3DFactory.ih [hunk 7]
```diff
@@ -94,9 +94,9 @@ DGtal::Display3DFactory<Space, KSpace>::draw(Display3D<Space, KSpace>& display,
         }
     }
 
-    typedef typename DGtal::GradientColorMap<Scalar, DGtal::CMAP_JET> ColorMap;
-    const ColorMap colormap(cmap_min, cmap_max);
-
+    typedef typename DGtal::HueShadeColorMap<Scalar, 2 > ColorMap;
+    const ColorMap colormap((cmap_min), (cmap_max));
```
Comment (kerautret) [inline src/DGtal/io/Display3DFactory.ih]: parenthesis ?

File: src/DGtal/io/Display3DFactory.ih [hunk 8]
```diff
@@ -109,8 +109,10 @@ DGtal::Display3DFactory<Space, KSpace>::draw(Display3D<Space, KSpace>& display,
         Scalar displayed_value = kform.myContainer(index);
         if (flipped) displayed_value = -displayed_value;
 
+        display << SetMode3D(cell.className(), "Basic");
+
         if (std::isfinite(displayed_value))
-            display << DGtal::CustomColors3D(DGtal::Color::Black, colormap(displayed_value));
+          display << DGtal::CustomColors3D(DGtal::Color::Black, colormap((displayed_value)) );
```
Comment (kerautret) [inline src/DGtal/io/Display3DFactory.ih]: also here ?

**Comment by kerautret** [comment]
looks fine, just small comments

File: src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+      ~SphereFittingEstimator( )
+      {
+        delete myFit ;
+      }
+
+      
+      /**
+       * Add the geometrical embedding of a surfel to the point list
+       *
+       * @param aSurf a surfel to add
```
Comment (kerautret) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: precise the [in] with@param  ?

File: ChangeLog.md (hunk 1 above)
Comment (dcoeurjo) [inline ChangeLog.md]: fixed thx

File: src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h (hunk 4 above)
Comment (dcoeurjo) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: done

File: src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h (hunk 5 above)
Comment (dcoeurjo) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: I would say no: this weight function comes from Mellado and I'd prefer to keep this setting from Mellado's documentation. I would agree if you could use DGtal weight functions but it is not the case.

File: src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h (hunk 6 above)
Comment (dcoeurjo) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: agree.

Comment (dcoeurjo) [inline src/DGtal/geometry/surfaces/estimation/estimationFunctors/SphereFittingEstimator.h]: Done (using const pointer, I've managed to do it with class static const...)

**Comment by dcoeurjo** [comment]
Done, thanks

**Comment by kerautret** [comment]
perfect merging...

**Pull request merged**